tensorflow/cc/tools/freeze_saved_model.cc

```diff
@@ -171,6 +172,47 @@ void ConvertReadVariableOpToIdentity(const NodeDef& node,
   identity_node->add_input(node.input(0));
 }
 
+// Returns the name of the VarHandleOp that provides input (possibly indirectly)
+// to node with node_name. A typical indirect chain of nodes (that can occur due
+// to graph inlining) is the following: VarHandleOp -> Identity -> Identity ->
+// ReadVariableOp. Calling the function on any of these nodes would return the
+// name of the VarHandleOp.
+StatusOr<string> GetVarHandleName(
+    const std::unordered_map<string, NodeDef*>& name_to_node_map,
+    string node_name) {
+  const NodeDef* node = name_to_node_map.at(node_name);
+  while (node->input_size() > 0) {
+    auto parent = name_to_node_map.find(node->input(0));
+    if (parent == name_to_node_map.end()) break;
+    node = parent->second;
+    if (node->op() != "Identity") {
+      VLOG(2) << "Stopping at non-identity node " << node->op();
+      break;
+    }
+  }
+  if (node->op() == "VarHandleOp") {
+    return node->name();
+  }
+  return errors::NotFound("No VarHandleOp ancestor found");
+}
+
+// Returns true if node_name has input from a VarHandleOp and that variable
+// handle represents a variable that we want to freeze (i.e. the handle name
+// is contained in variable_node_names). The name of the VarHandleOp is
+// returned in handle_name.
+bool HasVarHandleInput(
+    const std::unordered_map<string, NodeDef*>& name_to_node_map,
+    string node_name, const std::unordered_set<string>& variable_node_names,
+    string* handle_name) {
```

Comment:
Can we make the function to return an optional string instead?

Reply:
I have changed it to return `StatusOr<string>`, since that type was already used.